### PR TITLE
SR-IOV support

### PR DESCRIPTION
**Pull request by sil2100**
## Description

This might still need some work, but at least the base support is here.

This adds support for SR-IOV devices. One can now define VFs for given PFs via a 'link:' tag. Netplan then will create those dynamically on netplan apply. We also have a quick framework in case some special devices need a module unload+load to get the changes deployed.

This also adds support for SR-IOV hardware VLAN tagging via the renderer: sriov syntax.

I might need to write some integration tests for this if possible, but first I want to finish real-hardware testing.

## Checklist

- [x] Runs `make check` successfully.
- [x] Retains 100% code coverage (`make check-coverage`).
- [x] New/changed keys in YAML format are documented.
- [ ] \(Optional\) Closes an open bug in Launchpad.

### CODE REVIEW

**Comment by sil2100** [comment]
Open question for this PR: do we need to support matching of multiple PFs and VFs via the 'match:' syntax? Right now we do to some extent, although this complicates the whole code a lot. I think this actually doesn't make much sense, so I'll probably rip those pieces out.

File: src/parse.h [hunk 1]
```diff
@@ -314,6 +314,11 @@ struct net_definition {
     NetplanAuthenticationSettings auth;
     gboolean has_auth;
 
+    /* these properties are only valid for SR-IOV NICs */
+    struct net_definition* pf_link;
+    struct net_definition* vf_vlan_link;
```
[the remark this answers is not in the document]
Comment (sil2100) [reply src/parse.h:319]: Ah, indeed! I think this was some leftover from when I actually started working on this. Good catch, removing!

File: tests/generator/test_vlans.py [hunk 2]
```diff
@@ -105,6 +105,52 @@ def test_vlan(self):  # pragma: nocover
 unmanaged-devices+=interface-name:en1,interface-name:enblue,interface-name:enred,interface-name:engreen,''')
         self.assert_nm_udev(None)
 
+    @unittest.skipIf("CODECOV_TOKEN" in os.environ, "Skipping on codecov.io: GLib changed hashtable elements order")
```
[the remark this answers is not in the document]
Comment (sil2100) [reply tests/generator/test_vlans.py:108]: Another good catch, guess I copied it over when I copied the original test method.

File: netplan/cli/sriov.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+        with open(numvfs_path) as f:
+            vf_current = int(f.read().strip())
+        with open(totalvfs_path) as f:
+            vf_max = int(f.read().strip())
+    except IOError as e:
+        raise RuntimeError('failed parsing sriov_numvfs/sriov_totalvfs for %s: %s' % (pf, str(e)))
+    except ValueError:
+        raise RuntimeError('invalid sriov_numvfs/sriov_totalvfs value for %s' % pf)
+
+    if vf_count > vf_max:
```
[the remark this answers is not in the document]
Comment (sil2100) [reply netplan/cli/sriov.py:92]: Sounds like a sane proposition, +1 on that.

File: netplan/cli/sriov.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+            device_id = f.read().strip()[2:]
+        with open(os.path.join(devdir, 'device')) as f:
+            vendor_id = f.read().strip()[2:]
+    except IOError as e:
+        raise RuntimeError('could not determine vendor and device ID of %s: %s' % (pf, str(e)))
+
+    combined_id = ':'.join([vendor_id, device_id])
+    quirk_devices = ()  # TODO: add entries to the list
+    if combined_id in quirk_devices:
+        # some devices need special handling, so this is the place
```
[the remark this answers is not in the document]
Comment (sil2100) [reply netplan/cli/sriov.py:141]: So I didn't actually look for it too much, but I added this as a precaution as Stephane Graber and a few other people, who had a lot of experience with SR-IOV devices, said that many Intel and Mellanox ones required a kernel module reload for the actual new sriov_numvfs value to get accepted. That being said, on the SR-IOV Intel card that I have been given access to everything works as expected (no quirks needed). Same for documentation - both the Mellanox and Intel documentation on SR-IOV handling on Linux do not mention anything about any issues like that. I assume this might be the case for some cheaper cards perhaps?
Anyway, I will try getting some more info on that. I'll at least try to explain the situation there better in comments.

File: netplan/cli/sriov.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+                break
+
+    if not vf_index:
+        raise RuntimeError(
+            'could not determine the VF index for %s while configuring vlan %s' % (vf, vlan_name))
+
+    # now, create the VLAN filter
+    # ip link set dev PF vf ID vlan VID
+    try:
+        subprocess.check_call(['ip', 'link', 'set',
```
[the remark this answers is not in the document]
Comment (sil2100) [reply netplan/cli/sriov.py:169]: Ah, indeed we could! There's a lot of places in netplan where we could just use python directly. Let me add a TODO and let's do that as part of one refactoring branch.

File: src/parse.c [hunk 6]
```diff
@@ -667,6 +667,13 @@ parse_renderer(yaml_node_t* node, NetplanBackend* backend, GError** error)
 static gboolean
 handle_netdef_renderer(yaml_document_t* doc, yaml_node_t* node, const void* _, GError** error)
 {
+    if (cur_netdef->type == NETPLAN_DEF_TYPE_VLAN) {
+        if (strcmp(scalar(node), "sriov") == 0) {
+            cur_netdef->vf_filter = TRUE;
```
[the remark this answers is not in the document]
Comment (sil2100) [reply src/parse.c:672]: +1 on the rename, I think it makes more sense.

As for removing vf_filter - I don't think that's possible. The main purpose of the vf_filter variable (which I guess I'll rename to sriov_vlan_filter maybe!) is to mark the given vlan definition to be handled directly via the netlink layer. So for SR-IOV VF devices it's possible to have both regular VLANs and the so-called hardware filtered VLANs. A VF device, as any device, can have any number of vlans, but can also have a special hardware one set if the user uses the `renderer: sriov` tag it its definition. Those can only be one-per-VF.

So even though that would look better, it would mean that VFs could then not have regular vlans, which we do not want. At least if I understand your idea correctly!

[the remark this answers is not in the document]
Comment (slyon) [reply src/parse.c:672]: Oh, I see. I wasn't aware of the possibility to have HW and SW defined VLANs on those devices. So let's keep it like this.

**Comment by sil2100** [comment]
Thank you for your review! I also pushed a quick fix for an issue I have noticed during testing on real hardware. I didn't yet fully test the hardware VLAN filtering, but other than that this seems to work to some extent (and not crash!). So I think it might be good for merging.

Also, in the meantime, I started working on a feature-request in SR-IOV - the ability to explicitly state the number of VFs for a PF (additionally to us getting the number from the number of VFs configured).
The work on that can be seen here: https://github.com/CanonicalLtd/netplan/tree/sriov-support-explicit

It's based on this branch. Let's think about it, would welcome some more comments on both branches. Thank you!

**Comment by slyon** [comment]
This PR looks good for merging to me!

Regarding the `sriov-support-explicit` branch: So adding/removing VFs is not as invasive as expected? I like it setting set the real number of VFs declared in the netdef, this makes it more explicit and understandable: X VFs declared in netdef => Number of actual VFs configured for the PF = X.

But what is the usecase for setting `sriov_explicit_vf_count` > VFs declared in the netdef? Can you explain this a little more? As far as I understand, some VFs might go unused in this case... (As before when VFs were pre-set to some number bigger than the declared VFs from the netdef.)

**Comment by sil2100** [comment]
Hey! Thank you for your review! I'll do one more test on real hardware for the hardware VLAN support and then merge it - and propose the other branch as a separate PR.

Regarding that branch: yeah, so I actually also really really like the idea of what we have right now: number of VFs being equivalent to the number of VFs defined. It's clean and well defined. That being said, I have been reached out to by James Page asking if there could be a way of pre-defining the number of VFs to allocate. This apparently would be very useful for OpenStack, where none of the VFs would be configurable via netplan but could be allocated by it. Which is why I thought that maybe we could just add an 'optional' way of pre-allocating VFs for a given PF. We would then advertise it as being optional and only to be used in special cases.
And yes, I'm still hearing from people that allocating new VFs is very invasive, as frequently requiring the driver module reload in the process - though I have not yet experienced it myself. This might also help in such cases then, where administrators could allocate more VFs 'just in case'.

**Comment by sil2100** [comment]
Ok, VLAN setup looking good. Due to the lack of time I didn't test all the corner cases, but we can deal with those when they pop up.
